File: core/src/main/java/com/wamteavm/interpolator/interpfunction/LinearInterpolationFunction.java
package com.wamteavm.interpolator.interpfunction;

public class LinearInterpolationFunction<I extends Number> extends InterpolationFunction<I, Double> {

    public LinearInterpolationFunction(I[] x, Double[] y) {
        super(x, y);
    }

    @Override
    public Double evaluate(I at) {
        double atDouble = at.doubleValue();
        I[] genericI = getI();
        double[] i = new double[genericI.length];
        for (int index = 0; index < i.length; index++) {
            i[index] = genericI[index].doubleValue();
        }
        Double[] y = getO();
        if (atDouble <= i[0]) {
            return y[0];
        }
        if (atDouble >= i[i.length - 1]) {
            return y[y.length - 1];
        }

        int index = 0;
        while (atDouble > i[index+1]) {
            index++;
        }

        double x0 = i[index];
        double x1 = i[index+1];
        double y0 = y[index];
        double y1 = y[index+1];

        return y0 + (y1 - y0) * (atDouble - x0) / (x1 - x0);
    }

    @Override
    public void init() {

    }
}

File: core/src/main/java/com/wamteavm/interpolator/interpfunction/LinearInterpolationFunction.kt
package com.wamteavm.interpolator.interpfunction

class LinearInterpolationFunction<I : Number>(x: Array<I>, y: DoubleArray) : InterpolationFunction<I, Double>(x, y.toTypedArray()) {
    override fun evaluate(at: I): Double {
        val atDouble = at.toDouble()

        if (atDouble <= i[0].toDouble()) {
            return o[0]
        }
        if (atDouble >= i[i.size - 1].toDouble()) {
            return o[o.size - 1]
        }

        var index = 0
        while (atDouble > i[index + 1].toDouble()) {
            index++
        }

        val x0 = i[index].toDouble()
        val x1 = i[index + 1].toDouble()
        val y0 = o[index]
        val y1 = o[index + 1]

        return y0 + (y1 - y0) * (atDouble - x0) / (x1 - x0)
    }

    override fun init() {
    }
}
